consensus/XDPoS/engines/engine_v1/engine.go

```diff
@@ -727,14 +727,16 @@ func (x *XDPoS_v1) Prepare(chain consensus.ChainReader, header *types.Header) er
 				copy(header.Nonce[:], utils.NonceDropVote)
 			}
 		}
-		x.lock.RUnlock()
 	}
+	signer := x.signer
+	x.lock.RUnlock()
+
 	parent := chain.GetHeader(header.ParentHash, number-1)
 	if parent == nil {
 		return consensus.ErrUnknownAncestor
 	}
 	// Set the correct difficulty
-	header.Difficulty = x.calcDifficulty(chain, parent, x.signer)
+	header.Difficulty = x.calcDifficulty(chain, parent, signer)
```

Comment:
How about this version ?

```go
	if number%x.config.Epoch != 0 {
		// Gather all the proposals that make sense voting on
		addresses := make([]common.Address, 0, len(x.proposals))
		for address, authorize := range x.proposals {
			if snap.validVote(address, authorize) {
				addresses = append(addresses, address)
			}
		}
		// If there's pending proposals, cast a vote on them
		if len(addresses) > 0 {
			header.Coinbase = addresses[rand.Intn(len(addresses))]
			if x.proposals[header.Coinbase] {
				copy(header.Nonce[:], utils.NonceAuthVote)
			} else {
				copy(header.Nonce[:], utils.NonceDropVote)
			}
		}
	}

	parent := chain.GetHeader(header.ParentHash, number-1)
	if parent == nil {
		return consensus.ErrUnknownAncestor
	}
	// Set the correct difficulty
	x.lock.RLock()
	signer := x.signer
	x.lock.RUnlock()
	header.Difficulty = x.calcDifficulty(chain, parent, signer)
```

Reply:
The locks are already in the code before, i'm not sure what values it's locking, probably x.proposals. I prefer not to change the original logic as it would involve unnecessary riskks

Reply:
You can refer the struct and other codes:
```go
lock   sync.RWMutex    // Protects the signer fields
```

Reply:
I'm not sure i understand what you mean above. 
What's i'm referring to is that the lock is already within the original if statement. 
<img width="638" alt="image" src="https://github.com/XinFinOrg/XDPoSChain/assets/10041790/cc1b4430-5394-4ba7-b00d-ecb73d073b98">

I do not intend to change the behaviour of the original code hence move the `signer := x.signer` right before the Runlock in the original code. This would be the smallest change that would be required to achieve the same goal

Reply:
The `lock` field protects signer field.